### PR TITLE
Added write permissions feature to Dir section

### LINKED ISSUE #75: Add "write permission" feature to dir section
## Feature Request

**Is your feature request related to a problem? Please describe.**
This feature is intended to add the feature enhancement added to the dir section in the following two PRs:
https://github.com/denysdovhan/spaceship-prompt/pull/340
https://github.com/denysdovhan/spaceship-prompt/pull/517

**Describe the solution you'd like**
Porting over the existing zsh functionality of displaying write permissions in the dir section file that was added in this commit: https://github.com/denysdovhan/spaceship-prompt/commit/a5d5d31aee54a00833b1f7114793290038d4e00f

### PR DESCRIPTION
**Pull request by jskrnbindra**
Implement write permissions feature in the Dir section which shows a lock symbol (🔒) when:
- The current directory is write protected; or
- The current user does not have permissions enough to write to the directory

**Note:** Requires Powerline patched fonts to be installed.

## Motivation and Context

Closes #75 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Screenshots (if appropriate):
screen shot 2018-10-25 at 20 26 32

## How Has This Been Tested?

- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**

## Checklist:

- [x] I have checked that no other PR duplicates mine
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: functions/__sf_section_dir.fish
Comment (jskrnbindra) [inline functions/__sf_section_dir.fish:20]: @matchai  @Snuggle  How do we set the color of SPACEFISH_DIR_LOCK_SYMBOL?
Because as in [spaceship](https://github.com/denysdovhan/spaceship-prompt/commit/a5d5d31aee54a00833b1f7114793290038d4e00f)  the ```zsh``` way is easier.
Our ```__sf_lib_section``` seems to be hindering this implementation.
Correct me if I'm wrong.

**Comment by jskrnbindra** [comment]
@matchai A few tests are failing in AppVeyor but pass in my local machine.
Is this because I'm doing ```chmod``` anyhow? 🤔

File: functions/__sf_section_dir.fish
Comment (matchai) [reply functions/__sf_section_dir.fish:20]: @jskrnbindra Thanks for getting started on this PR! 😄 

Spaceship appears to circumvent the issue by having the lock symbol be displayed within the suffix using its color code, which I believe should work for spacefish as well.

Similarly to their implementation [here](https://github.com/denysdovhan/spaceship-prompt/commit/a5d5d31aee54a00833b1f7114793290038d4e00f#diff-4750536835f9cdbc54eafc032ad2a9a3R60), the following might work:

```fish
set SPACEFISH_DIR_SUFFIX (set_color $SPACEFISH_DIR_LOCK_COLOR)$SPACEFISH_DIR_LOCK_SYMBOL\
(set_color --bold fff)$SPACESHIP_DIR_SUFFIX
```

**Comment by matchai** [comment]
@jskrnbindra Hmm... Strange indeed. 🤔 
Could it be because of Cygwin user groups?
https://stackoverflow.com/a/19779410

**Comment by Snuggle** [comment]
Just going off your screenshot, are there _two_ spaces here or is it just your terminal font having weird kerning for emoji?

image

File: functions/__sf_section_dir.fish
Comment (jskrnbindra) [reply functions/__sf_section_dir.fish:20]: Cool. Worked nicely!

**Comment by jskrnbindra** [comment]
@matchai About the AppVeyor's chmod thing.
I tried: 
```bash
chown :Users testDir
```

and:
```bash
setfacl -s user::r-x,group::---,other::--- testDir
```
from [here](https://superuser.com/a/876913) as AppVeyor is using Cygwin v2.889.
If I try digging further it adds dirty commits to the PR. (Need two remove all commits after f07ae94)

The issue seems to be with Cygwin because the TavisCI build passes smoothly.
Also, why are we using two CI systems? 🤔
Can we make either of the builds passing as accepted (i.e. not forcing both builds to pass, if any of these passes it's okay)?

**Comment by jskrnbindra** [comment]
> Just going off your screenshot, are there _two_ spaces here or is it just your terminal font having weird kerning for emoji?
> 
> image

No these are not two spaces. It's my terminal displaying this Powerline font as such:
screen shot 2018-10-27 at 09 13 38
The lock emoji in the screenshot above is just the emoji between the quotes (and no space). The space after the lock symbol is within this emoji.

Is there a fix/workaround for this issue?
@Snuggle

**Comment by matchai** [comment]
The reason we have two CI systems is to cover all supported OSes. Travis tests Linux and macOS while AppVeyor tests Windows. 🍎🐧 

Don't worry about dirtying the commit history. We squash-merge, so this PR will be recorded as one commit. 👍 

With regards to your terminal showing extra space, I personally use iTerm 2. Under Profiles > Text > Unicode, you can enable *Use Unicode version 9 widths*.

File: functions/__sf_section_dir.fish
Comment (matchai) [inline functions/__sf_section_dir.fish]: It appears spaceship uses the powerline font's lock as opposed to the emoji lock `" "`
I would suggest doing the same to maintain consistency.
https://github.com/denysdovhan/spaceship-prompt/blob/0f4f6840ffba60c1f8b3e329ff2ee5a69d48a983/sections/dir.zsh#L17

Comment (jskrnbindra) [reply functions/__sf_section_dir.fish]: I used Powerline lock symbol from the Character Viewer app in mac. Didn't work.
Now I've just copied from spaceship. It works.

**Comment by jskrnbindra** [comment]
@matchai I tried a lot of ways. (```chgrp```, ```chown```, ```setfacl```, etc.)
But AppVeyor's filesystem (NTFS) does not seems to make a directory write protected.
I tried modifying users, owners and groups but no luck.

A simple ```chmod``` was able to update the directory permissions as expected but only superficially. For example, running
```bash 
chmod 400 testDir
```
or 
```bash 
chmod -w testDir
```
would show:
```dr-x------+ 1 appveyor Users 0 Oct 27 09:44 testDir```
as output of ```ls -lahrt```.
But when I ```cd``` into testDir and try to create a file, it creates it.
and even ```echo ([ -w . ]; and echo "writable")``` ran inside testDir (after running the above ```chmod```) prints 'writable' to the screen.

For Windows (Cygwin) we need to find out a way to:
-  make the dir read-only
- see if a directory is read-only.

Cygwin docs suggest using ```setfacl``` command line tool to change permissions for files but I was unable to run this command in AppVeyor's environment.

File: docs/Options.md
Comment (salmanulfarzy) [inline docs/Options.md]: It's better not to add images to repository because they're cloned during installation. Instead we could upload it to comments and then use that link in the doc.

Comment (jskrnbindra) [reply docs/Options.md]: 46248218-4af95d80-c434-11e8-8e25-595d792503f1
🎉

**Comment by jskrnbindra** [comment]
I think we should put a **"help-wanted"** label on this.

**Comment by jskrnbindra** [comment]
Should we use some hacky way of doing this?
Like looking at the output of ```ls -lahrt``` to see if the directory is write protected or not?

Or should we drop this feature off on Windows for now?

**Comment by matchai** [comment]
I'd like to do a little bit of testing on my personal Cygwin machine to see if I could maybe get this working before we decide to have features only enabled for some terminal emulators.

I'll try to get back to this PR in the next few days after I'm done traveling ✈️

**Comment by Snuggle** [comment]
Why Cygwin? Isn't WSL the standard for running Linux stuff on Windows now?

**Comment by matchai** [comment]
I'm personally running Windows 10 LTSB, which has no support yet for WSL yet; so a little bit for selfish reasons 😉.

EDIT: Whoops! Looks like I'm wrong: https://docs.microsoft.com/en-us/windows/wsl/install-manual
It would still be nice to keep Cygwin supported for users that aren't on Win10, but if it's a headache, I suppose we can drop those tests.

**Comment by Snuggle** [comment]
Windows Subsystem for Linux seems much much nicer. I just don't know of a way to test for it, do you?

**Comment by jskrnbindra** [comment]
Considering WSL, testing shouldn't be any different from testing on Linux. So if a test passes on linux it will pass in WSL.

**Comment by Snuggle** [comment]
In theory, yup! WSL seems much more important than Cygwin though.

**Comment by matchai** [comment]
I don't believe that any CI providers at the moment support nested virtualization for free for OSS: https://github.com/appveyor/ci/issues/1295#issuecomment-412597732

**Comment by kulabun** [comment]
I think that since it is a cygwin bug, the best way is just to ignore cygwin.
```
if test (uname -s) != "CYGWIN"
   echo $SPACEFISH_DIR_LOCK_SYMBOL
end
```

**Comment by matchai** [comment]
I've been holding up this PR for too long.
I suppose it's fine if we ignore Cygwin when testing this section for the time being. @jskrnbindra if you can get this to a passing state, I'll go ahead and give it another look over. 👍

**Comment by matchai** [comment]
It passes! 🙌 
Thank you for taking the time to get this working @jskrnbindra 😄 
You may just want to rebase master since your branch is out of date.

I'll get started with reviewing it.

File: functions/__sf_section_dir.fish
Comment (matchai) [inline functions/__sf_section_dir.fish]: It looks like we should be able to join the two conditionals here. 😄
```suggestion
    if [ $SPACEFISH_DIR_LOCK_SHOW = true -a ! -w . ];
```

File: tests/__sf_section_dir.test.fish
Comment (matchai) [inline tests/__sf_section_dir.test.fish]: It looks like we have some mixed whitespace over here. 🤔

File: tests/__sf_section_dir.test.fish
Comment (matchai) [inline tests/__sf_section_dir.test.fish:6]: To slim down on the setup in each test to do with write protected directories, I'd suggest adding this to the setup so you can just `cd` into them at the start of the test:
```suggestion
	mkdir -p /tmp/tmp-spacefish/dir1/dir2/dir3
	mkdir -p /tmp/tmp-spacefish/writeProtected
	chmod 500 /tmp/tmp-spacefish/writeProtected
```

File: tests/__sf_section_dir.test.fish
Comment (matchai) [inline tests/__sf_section_dir.test.fish]: Oops! Looks like this test wasn't written yet 😮

File: tests/__sf_section_dir.test.fish
Comment (matchai) [inline tests/__sf_section_dir.test.fish:347]: Oops! Looks like this test wasn't written yet 😮

Comment (jskrnbindra) [reply tests/__sf_section_dir.test.fish:347]: oh! 🙈

File: docs/Options.md
Comment (matchai) [inline docs/Options.md]: Unrelated to your changes, but I just noticed this:
```suggestion
| `SPACEFISH_DIR_COLOR` | `cyan` | Color of directory section |
```

**Review by matchai — APPROVED** [review]
Looks good to me! 👍 
If you don't mind fixing that documentation issue from before, I'll go ahead and merge this in.

**Comment by jskrnbindra** [comment]
🎉🙃

File: docs/Options.md
Comment (Snuggle) [inline docs/Options.md:72]:
```suggestion
Directory is always shown and truncated to the value of `SPACEFISH_DIR_TRUNC`. While you are in Git repository, this section shows only the project title and relevant status icons. If current directory is write-protected or if current user doesn't have write permissions, a padlock (by default) will be displayed as a suffix.
```

**Comment by matchai** [comment]
Thanks again @jskrnbindra!
I know this took a bit of a while, but I'm glad you stuck with it. 😄

File: docs/Options.md
Comment (Snuggle) [inline docs/Options.md]: Why is there an image in here? 😕

File: functions/__sf_section_dir.fish
Comment (Snuggle) [inline functions/__sf_section_dir.fish:44]: Indentation is borked here. If/end statements are having a dance!

File: tests/__sf_section_dir.test.fish
Comment (Snuggle) [inline tests/__sf_section_dir.test.fish:11]: This line is randomly indented – must be going to the store! :yum:

**Review by Snuggle — COMMENTED** [review]
I've made some suggestions for code and documentation clean-up!
Feel free to add these changes and/or to add some discussion below, I'd love to hear what you think. :sparkling_heart: :sparkles:

**Comment by Snuggle** [comment]
Oh gosh, this PR got merged seconds before I managed to get my review in time, heh. 😅
Battle plan?

File: functions/__sf_section_dir.fish
Comment (Snuggle) [reply functions/__sf_section_dir.fish:44]: (Not visible from GitHub's preview, you'd have to view the whole file to see the `end` statement.)
Line `43` and Line `45` have different indentation levels.

File: docs/Options.md
Comment (Snuggle) [reply docs/Options.md:72]: This paragraph of text doesn't make much sense and could be quite confusing, here's a quick syntactic clean-up.

**Comment by matchai** [comment]
Whoops! I'll quickly commit your suggested changes to master. 😅

**Comment by jskrnbindra** [comment]
Lost all the permissions to modify on this PR now. ⚓️

**Pull request merged**
